src/ser.rs

```diff
@@ -429,14 +429,50 @@ pub struct CollectMap<'a, W> {
     ser: &'a mut Serializer<W>,
 }
 
-impl<'a, W> CollectMap<'a, W> {
+impl<'a, W> CollectMap<'a, W>
+where
+    W: enc::Write,
+{
     fn new(ser: &'a mut Serializer<W>) -> Self {
         Self {
             buffer: BufWriter::new(Vec::new()),
             entries: Vec::new(),
             ser,
         }
     }
+
+    fn serialize<T: Serialize + ?Sized>(
+        &mut self,
+        maybe_key: Option<&'static str>,
+        value: &T,
+    ) -> Result<(), EncodeError<W::Error>> {
+        // Instantiate a new serializer, so that the buffer can be re-used.
+        let mut mem_serializer = Serializer::new(&mut self.buffer);
+        if let Some(key) = maybe_key {
+            key.serialize(&mut mem_serializer)
+                .map_err(|_| EncodeError::Msg("Struct key cannot be serialized.".to_string()))?;
+        }
+        value
+            .serialize(&mut mem_serializer)
+            .map_err(|_| EncodeError::Msg("Struct value cannot be serialized.".to_string()))?;
+
+        self.entries.push(self.buffer.buffer().to_vec());
+        self.buffer.clear();
+
+        Ok(())
+    }
+
+    fn end(mut self) -> Result<(), EncodeError<W::Error>> {
+        // This sorting step makes sure we have the expected order of the keys. Byte-wise
+        // comparison gives us the right order as keys in DAG-CBOR are always (text) strings, hence
+        // have the same CBOR major type 3. The length of the string is encoded in the following
+        // bits. This means that a shorter string sorts before a longer string.
+        self.entries.sort_unstable();
```

Comment:
I don't think this is enough. We're stuck with the 7049 rules of length-first sorting. https://ipld.io/specs/codecs/dag-cbor/spec/#strictness

A good extension to your tests would be to add a new field to the struct "a1" which should come after "b".

Reply:
Thanks for pointing out and proposing better tests. Though I think it'ss correct. In a previous PR Steb asked me to clarify this a bit in a comment, but it still doesn't seem to be clear. When you've an idea to word it in a better way, please let me know.

As CBOR prefixes the strings with the length, shorter strings are sorted first. I even think this is the reason they came up with this (counter-intuitive) ordering. So that you can take the full encoded value (not just the string itself) and memcmp it for sorting.

Reply:
Importantly, even though the length itself is variable length, the length itself is either:

1. 0-23: inlined into the 5-bit additional data field.
2. 24-27: 1-8 byte length.

Which means that lexicographical sorting actually works.

Reply:
Oh, right, I missed that this was sorting over _encoded_ values; and yeah, neat that it actually works given the flexible prefix!